### ISSUE TITLE
How do I get Cilium to run on WSL2?

### ISSUE THREAD

**Issue by HummingMind** [cilium/cilium#29302]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Cilium pods crashing the the following error message in version 1.15.0-pre.2:

level=fatal msg="failed to start: daemon creation failed: error while initializing daemon: failed while reinitializing datapath: removing ipv6 proxy routing rule: address family not supported by protocol" subsys=daemon

Version 1.14.4 works fine.

### Cilium Version

1.15.0-pre.2

### Kernel Version

linux-msft-wsl-5.15.133.1

### Kubernetes Version

1.28.3

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by HummingMind** [comment]
I understand this might be related to ipv6, but what changed between 1.14.4 and 1.15.0-pre.2? Works fine with 1.14.4 but not with 1.15.0-pre.2

WSL2, kind cluster, with custom-compiled kernel with all the required modules according to: 
https://docs.cilium.io/en/v1.12/operations/system_requirements/

**Comment by networkop** [comment]
I believe `CONFIG_IPV6_MULTIPLE_TABLES=y` may be missing

**Comment by ti-mo** [comment]
> but what changed between 1.14.4 and 1.15.0-pre.2?

tl;dr: a lot. These rules used to be installed/removed by a shell script with a bunch of `ip rule .. || true` that swept all these errors under the rug. Fortunately, this is no longer the case. :)

@HummingMind I've put up https://github.com/cilium/cilium/pull/29311 that ignores only the specific EAFNOSUPPORT returned by the rule removal. Feel free to play around with this on your own. Are WSL2 kernels typically built with v6 disabled? In this case, you'll need to disable Cilium's v6 support. Otherwise, you'll just hit errors elsewhere.

@networkop's suggestion above may also prove useful. Did you explicitly disable `CONFIG_IPV6_MULTIPLE_TABLES`? I can't figure out from the kernel docs whether this defaults to true or not, but it's pulled in by at least CONFIG_IPV6_SEG6_LWTUNNEL. Unfortunately, there are many places returning EAFNOSUPPORT.

**Comment by HummingMind** [comment]
> I believe `CONFIG_IPV6_MULTIPLE_TABLES=y` may be missing

I just checked the config-wsl file, and yes, this is not set.

**Comment by HummingMind** [comment]
> Are WSL2 kernels typically built with v6 disabled? In this case, you'll need to disable Cilium's v6 support. Otherwise, you'll just hit errors elsewhere.

I believe so. There is a new netwroking mode in WSL2 2.0.9, called mirrored mode (experimental), which adds IPv6 support to WSL2. I'll try it out, and if it doesn't work, I'll go back to figuring out what might be missing from the compiled kernel (such as CONFIG_IPV6_MULTIPLE_TABLES=y that was mentioned by @networkop)


I wish you guys had official docs on compiling the WSL2 kernel, as it is a really popular dev environment for local kubernetes testing/development. 🙏🏻

**Comment by HummingMind** [comment]
> I can include a diff of non-default flags (1.15.0-pre.2 runs fine for me) that I've got enabled but that may include some of the stuff not needed by Cilium.

I'll play around with a few things, but if I can't get it to work, I'll take you up on the offer. Thank you!

**Comment by HummingMind** [comment]
WSL2 mirrored mode doesn't work with Docker yet. 🤣 
Issue here: https://github.com/microsoft/WSL/issues/10494

Will try disabling IPv6 during the install, to see if that works. If it doesn, I'll go back to messing with the WSL2 kernel. 😨

**Comment by HummingMind** [comment]
Looks like IPv6 is disabled by default anyway during the Cilium installation. I tried with --set ipv6.enabled=false flag just in case, but getting the same error.

**Comment by HummingMind** [comment]
> > I can include a diff of non-default flags (1.15.0-pre.2 runs fine for me) that I've got enabled but that may include some of the stuff not needed by Cilium.
> 
> I'll play around with a few things, but if I can't get it to work, I'll take you up on the offer. Thank you!

@networkop I think I'll take you up on your offer, since you have it working. Are you running v5 or v6 kernel? 🍻

**Comment by HummingMind** [comment]
I set CONFIG_IPV6_SEG6_LWTUNNEL to CONFIG_IPV6_SEG6_LWTUNNEL=y , and the original error message is gone.
New error is:

`level=info msg="Start hook executed" duration="10.68µs" function="*statedb.DB.Start" subsys=hive
level=error msg="Start hook failed" error="NewHandleAt failed: protocol not supported" function="*linux.devicesController.Start" subsys=hive
level=info msg=Stopping subsys=hive
level=info msg="Stop hook executed" duration="4.662µs" function="*statedb.DB.Stop" subsys=hive`

**Comment by HummingMind** [comment]
Got it to work! I just enabled most of the IPv6 modules in the networking section.

![image](https://github.com/cilium/cilium/assets/87399963/1aa99aae-f745-40db-98ec-0070987d94ce)

I guess you can close this!

Thank you both!

**Comment by ti-mo** [comment]
@HummingMind Thanks for reporting back! Would you be willing to contribute a small section to our documentation with your findings? Sounds like it would help out quite a few WSL2 users!

**Comment by HummingMind** [comment]
I can certainly contribute. Not sure I am qualified though, as this was the first time that I ever had to compile a Linux kernel. 😆

How would I go about doing this? I am also a bit new to open source 😨

**Comment by ti-mo** [comment]
> I can certainly contribute. Not sure I am qualified though, as this was the first time that I ever had to compile a Linux kernel. 😆

If you've made it through building your own kernel, you're officially qualified. 😉

> How would I go about doing this? I am also a bit new to open source 😨 

Not a problem! We have exhaustive documentation on the contribution process, see https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request. Granted, most of these things don't apply if you're just making a documentation change. Essentially, working on the docs implies editing a text file. Very straightforward! Get yourself a copy of the cilium/cilium source code and run `make render-docs` to get a live preview in your browser.

Now, I'm not sure what the scope of the docs should be here. We're not going to document the whole kernel build process, but maybe some general pointers would be nice. Note that we already document a set of minimal kernel configs needed for Cilium to function correctly. Maybe linking to https://wsl.dev/wslcilium could be useful to avoid documenting the whole process for now, at least until wsl2 gains better defaults or makes it easier to plug in custom kernels.

**Comment by sadath-12** [comment]
+1 adding this to docs would really help including me

**Comment by HummingMind** [comment]
Ok. I'll post the instruictions here in a bit. Someone else can submit them as a PR.

**Comment by sadath-12** [comment]
works thanks 😊

**Comment by sadath-12** [comment]
thanks going to follow and try this

**Comment by HummingMind** [comment]
This is more informational that anything else, but you can see here the desired configuration vs the default values:

![image](https://github.com/cilium/cilium/assets/87399963/421d86be-648a-4b51-b11f-25c6f76ab5fd)

This is for the options listed in https://docs.cilium.io/en/stable/operations/system_requirements/#linux-kernel

So does not include anything else that might be needed (TBD).

**Comment by HummingMind** [comment]
@ti-mo 

I was mistaken when I initially said: 

> I just enabled most of the IPv6 modules in the networking section.

Your documentation is correct and the modules you have listed in the docs are all what is needed.

I just did not configure them correctly the 1st time around, but after putting together the instructions for the post above, I found out that the docs are correct. The only catch was that these lines:

```
CONFIG_INET{,6}_ESP=m
CONFIG_INET{,6}_IPCOMP=m
CONFIG_INET{,6}_XFRM_TUNNEL=m
CONFIG_INET{,6}_TUNNEL=m
```

had to be entered as:

```
CONFIG_INET_ESP=m
CONFIG_INET_IPCOMP=m
CONFIG_INET_XFRM_TUNNEL=m
CONFIG_INET_TUNNEL=m
CONFIG_INET6_ESP=m
CONFIG_INET6_IPCOMP=m
CONFIG_INET6_XFRM_TUNNEL=m
CONFIG_INET6_TUNNEL=m
```

You can close this issue. The docs don't really need to be updated (unless people want to follow the steps as I outlined them in my posts above).

**Comment by sadath-12** [comment]
@HummingMind I did all those things but still when I do ```uname -r``` it shows the old kernel name whereas in configuration I had changed the name . also to check I installed cilium and face same issue again any reason why wsl might not be picking the new compiled kernel

**Comment by HummingMind** [comment]
@sadath-12 
uname -r should show 5.15.137.3-microsoft-standard-WSL2+ as your kernel name (or whatever you changed it to).

If it doesn't, that means you are not ponting at the new kernel in your wsl config.

Make sure that in Windows you create the ```.wslconfig``` file in your ```%USERPROFILE%``` location (would be something like C:\Users\yourusername). This file should have the path to your bzImage that is stored in your Windows storage somewhere:

```
[wsl2]
kernel=C:\\some_folder_on_your_windows_disk\\bzImage
```

Then, once you do ```wsl --shutdown```, you should be able to start your WSL2 instance again and it should return the new kernel version with ```uname -r```

**Comment by sadath-12** [comment]
yes the solution was to do it as ```code .wslconfig``` instead creating txt document it worked but when I open the wsl2 now I get ```The operation timed out because a response was not received from the virtual machine or container.
Error code: Wsl/Service/CreateInstance/CreateVm/HCS_E_CONNECTION_TIMEOUT
Press any key to continue...``` on the net for many people it works by ```wsl --shutdown``` does not seem to work for me

**Comment by sadath-12** [comment]
yes the solution was to do it as code .wslconfig instead creating txt document it worked but when I open the wsl2 now I get The operation timed out because a response was not received from the virtual machine or container. Error code: Wsl/Service/CreateInstance/CreateVm/HCS_E_CONNECTION_TIMEOUT Press any key to continue... on the net for many people it works by wsl --shutdown does not seem to work for me . later I referred to the issue https://github.com/microsoft/WSL/issues/10196 if I try something with single slash or put them under strings I revert back to old kernel

**Comment by sadath-12** [comment]
> This is more informational that anything else, but you can see here the desired configuration vs the default values:
> 
> ![image](https://private-user-images.githubusercontent.com/87399963/294614044-421d86be-648a-4b51-b11f-25c6f76ab5fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ2MDMwNjksIm5iZiI6MTcwNDYwMjc2OSwicGF0aCI6Ii84NzM5OTk2My8yOTQ2MTQwNDQtNDIxZDg2YmUtNjQ4YS00YjUxLWIxMWYtMjVjNmY3NmFiNWZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA3VDA0NDYwOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM0MmQ2MmMwYzIyNDZmMmJkZGFhMTRkMWM3N2JmMGU1YWZjZWE1MWE1Y2JiYWI1MjIxMjJiNDI0NzIxOTQ5MGYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3fb1zurrQMEEviqKzo0JmvR0LRZYA-Wu8RTSLpnr-a0)
> 
> This is for the options listed in https://docs.cilium.io/en/stable/operations/system_requirements/#linux-kernel
> 
> There are 42 options (as of Cilium 1.15.0) that need to be enabled and configured (wtih the desired values above).
> 
> Some were not set and commented out, some were missing, some just had Y instead of M (see the table for all the differences).
> 
> I use "grep" to find these lines (and determine what is missing) and "sed" to change the values or add the missing lines.

I did the configs from here and then for compiling and moving the kernel I referred https://wsl.dev/wslcilium/ and kept ```[wsl2]
kernel = C:\\wslkernel\\kernel-cilium``` in the .wslconfig

**Comment by sadath-12** [comment]
referring this https://github.com/microsoft/WSL/issues/8793#issuecomment-1365590699 I do also have it enabled -- ```C:\Windows\System32>bcdedit | find /i "hypervisor"
hypervisorlaunchtype    Auto```

**Comment by sadath-12** [comment]
ok I confirmed something is wrong with the kernel I compiled . Earlier before this discussion I used https://youtu.be/6lqMeg_n7l4?si=aH1M2gs94RDRzz5A but following the configuration you provided . wsl2 is able to locate this when I mention its path in .wslconfig but still I get those cilium problems

**Comment by sadath-12** [comment]
Currently I think since I used different ways to compile them it might be corrupted ill try the compilation from fresh again and try

**Comment by networkop** [comment]
I've also found that building the loadable modules works around a few quirks when Cilium tries to run `modprobe` in `FindOrLoadModules`. Without it, `FindOrLoadModules` returns an error and you Cilium tries to fall back to an alternative (e.g. from BPF-based masq to IPtables). 
Here are the instructions:

```
# set the kernel target kernel version (uname -r)
export KERNELRELEASE=5.15.90.3-microsoft-standard-WSL2+
# build the loadable modules
make KERNELRELEASE=$KERNELRELEASE -j16 KCONFIG_CONFIG=Microsoft/config-wsl modules
# copy modules to /lib/modules/
sudo make KERNELRELEASE=$KERNELRELEASE -j16 KCONFIG_CONFIG=Microsoft/config-wsl modules_install

```

**Comment by HummingMind** [comment]
> Currently I think since I used different ways to compile them it might be corrupted ill try the compilation from fresh again and try

Yeah, I would try a clean clone of the WSL2 kernel repo.

Also, I modified the last section of my original post to include the instructions on loading the modules, per @networkop recommendation (although he does it a bit differently).

**Comment by sadath-12** [comment]
> This is more informational that anything else, but you can see here the desired configuration vs the default values:
> 
> ![image](https://private-user-images.githubusercontent.com/87399963/294614044-421d86be-648a-4b51-b11f-25c6f76ab5fd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDQ3MDMxMDAsIm5iZiI6MTcwNDcwMjgwMCwicGF0aCI6Ii84NzM5OTk2My8yOTQ2MTQwNDQtNDIxZDg2YmUtNjQ4YS00YjUxLWIxMWYtMjVjNmY3NmFiNWZkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMDglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTA4VDA4MzMyMFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTcxOGVhNGFhYWQxMDliM2RjZmFkZWQ2OGQyYWM0ZTdlMGM0ZTlhZWIzYzczZWEwYmI0MzQ3M2FmOWM3YzA4YTkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Vm38zsPszEv5u3Pz4KbS2g15Trb1p5_ifmobodCCSQ4)
> 
> This is for the options listed in https://docs.cilium.io/en/stable/operations/system_requirements/#linux-kernel
> 
> There are 42 options (as of Cilium 1.15.0) that need to be enabled and configured (wtih the desired values above).
> 
> Some were not set and commented out, some were missing, some just had Y instead of M (see the table for all the differences).
> 
> I use "grep" to find these lines (and determine what is missing) and "sed" to change the values or add the missing lines.

@HummingMind  this is enough right ?

**Comment by sadath-12** [comment]
Now it worked easily for me . Thank you @HummingMind for making this thing possible . Really appreciate your every guide has been worth

**Comment by HummingMind** [comment]
@sadath-12 glad to hear you got it working! 🍻

**Comment by sadath-12** [comment]
@HummingMind  were you able t run tetragon by building images locally ? I can't seems there is a much bigger obstacle there for wsl2 users

**Comment by HummingMind** [comment]
> @HummingMind were you able t run tetragon by building images locally ? I can't seems there is a much bigger obstacle there for wsl2 users

I have not tried it.

**Comment by joestringer** [comment]
It seems like this question has run its course, and the tracking issue for WSL2 documentation is at https://github.com/cilium/cilium/issues/21542 . I'll close this one out.
